nixos/modules/system/boot/initrd-tailscale.nix

```diff
@@ -0,0 +1,79 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.boot.initrd.network.tailscale;
+
+in
+
+{
+
+  options = {
+
+    boot.initrd.network.tailscale.enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = lib.mdDoc ''
+        Starts an Tailscale client during initrd boot. It can be used to e.g.
+        remotely accessing the SSH service controlled by
+        {option}`boot.initrd.network.ssh` or other network services
+        included. Service is killed when stage-1 boot is finished.
+      '';
+    };
+
+    boot.initrd.network.tailscale.configuration = mkOption {
+      type = types.path; # Same type as boot.initrd.secrets
+      description = lib.mdDoc ''
+        The configuration file for OpenVPN.
+
+        ::: {.warning}
+        Unless your bootloader supports initrd secrets, this configuration
+        is stored insecurely in the global Nix store.
+        :::
+      '';
+      example = literalExpression "./configuration.ovpn";
+    };
+
+  };
+
+  config = mkIf (config.boot.initrd.network.enable && cfg.enable) {
+    assertions = [
+      {
+        assertion = cfg.configuration != null;
+        message = "You should specify a configuration for initrd OpenVPN";
+      }
+    ];
+
+    # Add kernel modules needed for Tailscale
+    boot.initrd.kernelModules = [ "tun" "tap" ];
+
+    # Add Tailscale and ip binaries to the initrd
+    # The shared libraries are required for DNS resolution
+    boot.initrd.extraUtilsCommands = ''
+      copy_bin_and_libs ${pkgs.tailscale}/bin/tailscale
+      copy_bin_and_libs ${pkgs.iproute2}/bin/ip
+
+      cp -pv ${pkgs.glibc}/lib/libresolv.so.2 $out/lib
```

Comment:
What are the min dependencies required by tailscale? Are these needed. Is Ip needed. etc.

Reply:
I have a (hacky and not quite ready for OSS-ing) Tailscale-in-initrd module working, and I'm using:
```nix
  let
      iptables-static = pkgs.iptables.overrideAttrs (old: {
        dontDisableStatic = true;
        configureFlags = (lib.remove "--enable-shared" old.configureFlags) ++ [
          "--enable-static"
          "--disable-shared"
        ];
      });
  in {
        boot.initrd.availableKernelModules = [
          "ip6_tables"
          "ip6t_rpfilter"
          "ip_tables"
          "ipt_rpfilter"
          "libcrc32c"
          "nf_conntrack"
          "nf_conntrack_netlink"
          "nf_defrag_ipv4"
          "nf_defrag_ipv6"
          "nf_nat"
          "nf_reject_ipv4"
          "nf_reject_ipv6"
          "nf_tables"
          "nft_chain_nat"
          "nft_compat"
          "nfnetlink"
          "nft_compat"
          "tun"
          "x_tables"
          "xt_LOG"
          "xt_MASQUERADE"
          "xt_addrtype"
          "xt_comment"
          "xt_conntrack"
          "xt_mark"
          "xt_multiport"
          "xt_pkttype"
          "xt_tcpudp"
        ];
        # [...]
        boot.initrd.extraUtilsCommands = ''
          copy_bin_and_libs ${pkgs.tailscale}/bin/.tailscaled-wrapped
          copy_bin_and_libs ${pkgs.tailscale}/bin/.tailscale-wrapped
          copy_bin_and_libs ${pkgs.iproute}/bin/ip
          copy_bin_and_libs ${iptables-static}/bin/iptables
          copy_bin_and_libs ${iptables-static}/bin/ip6tables
          copy_bin_and_libs ${iptables-static}/bin/xtables-legacy-multi
          copy_bin_and_libs ${iptables-static}/bin/xtables-nft-multi
          copy_bin_and_libs ${pkgs.strace}/bin/strace
        '';
  };
```

This may be slightly more than what is required, but still... it works!